### PR TITLE
HIVE-27439: Support Decimal(precision, scale) format

### PR DESCRIPTION
**Pull request by leesf**
### What changes were proposed in this pull request?
Now, TypeInfoUtils#parseType only support Decimal(precision,scale) instead of Decimal(precision, scale), for example, support decimal(10,2) instead of decimal(10, 2) with a blank before 2, however, users may create decimal in this format decimal(10, 2) and it will throw exception in 

```

Integer.parseInt(params[1]);

```

we need fix this issue.

### Why are the changes needed?
Fix the Decimal(precision, scale) parsing error.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

File: serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java
Comment (zhangbutao) [inline serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java:492]: Thanks for fixing this issue. IMO, we can fix it in upstream invoking code, then we can avoid other potential type problems caused by the  blank character. Like this:

```
diff --git a/serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java b/serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java
index b23e3cfc44..4788558d4b 100644
--- a/serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java
+++ b/serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java
@@ -334,7 +334,7 @@ private static boolean isTypeChar(char c) {
             || !isTypeChar(typeInfoString.charAt(end))) {
           Token t = new Token();
           t.position = begin;
-          t.text = typeInfoString.substring(begin, end);
+          t.text = typeInfoString.substring(begin, end).trim();
           t.isType = isTypeChar(typeInfoString.charAt(begin));
           tokens.add(t);
           begin = end;
```

Comment (leesf) [reply serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java]: done

**Review by zhangbutao — APPROVED** [review]
LGTM. +1 non-binding.

**Comment by leesf** [comment]
cc @nrg4878

**Comment by leesf** [comment]
cc @ayushtkn would you please take a look

File: serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java
Comment (ayushtkn) [reply serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java]: @zhangbutao Wasn't the previous approach more safe? If some token expects to have a space before or after, can we land up in a mess?

**Review by ayushtkn — APPROVED** [review]
LGTM, if the tests pass

**Comment by leesf** [comment]
> LGTM, if the tests pass

@ayushtkn The tests have already passed

**Comment by zhangbutao** [comment]
@leesf The linked JIRA is HIVE-27439  not HIVE-27321. Please change the PR description.

File: serde/src/test/org/apache/hadoop/hive/serde2/typeinfo/TestTypeInfoUtils.java
Comment (zhangbutao) [inline serde/src/test/org/apache/hadoop/hive/serde2/typeinfo/TestTypeInfoUtils.java:54]: HIVE-11476 added a special case that allowing spaces in some types. e.g `struct<user id:int,user group: int>` but it not added the test case. Could you please add this test case here to make sure we won't get a regression?

BTW, i have tested the case `struct<user id:int,user group: int>` with this change locally and it worked well. But i think it is necessary to add it explicitly.  cc @ayushtkn

**Comment by leesf** [comment]
> @leesf The linked JIRA is [HIVE-27439](https://issues.apache.org/jira/browse/HIVE-27439) not [HIVE-27321](https://issues.apache.org/jira/browse/HIVE-27321). Please change the PR description.

done

File: serde/src/test/org/apache/hadoop/hive/serde2/typeinfo/TestTypeInfoUtils.java
Comment (leesf) [reply serde/src/test/org/apache/hadoop/hive/serde2/typeinfo/TestTypeInfoUtils.java]: thanks, added the case.

**Pull request merged**
